hi-ml/testhiml/testhiml/test_run_ml.py

```diff
@@ -220,11 +220,41 @@ def test_run_training() -> None:
             mock_trainer.loggers[0].finalize.assert_called_once()
 
 
+@pytest.mark.parametrize("max_num_gpus_inf", [-1, 1])
+def test_end_training(max_num_gpus_inf: int) -> None:
+    experiment_config = ExperimentConfig(model="HelloWorld")
+    container = HelloWorld()
+    container.max_num_gpus_inference = max_num_gpus_inf
+    runner = MLRunner(experiment_config=experiment_config, container=container)
+
+    with patch.object(container, "get_data_module"):
+        with patch("health_ml.run_ml.create_lightning_trainer", return_value=(MagicMock(), MagicMock())):
+            runner.setup()
+            runner.init_training()
+            runner.run_training()
+
+        with patch.object(runner.checkpoint_handler, "additional_training_done") as mock_additional_training_done:
+            with patch.object(runner, "after_ddp_cleanup") as mock_after_ddp_cleanup:
+                with patch("health_ml.run_ml.cleanup_checkpoints") as mock_cleanup_ckpt:
+                    old_environ = {"old": "environ"}
+                    runner.end_training(old_environ=old_environ)
+                    mock_additional_training_done.assert_called_once()
+                    mock_cleanup_ckpt.assert_called_once()
+                    if max_num_gpus_inf == 1:
+                        mock_after_ddp_cleanup.assert_called_once()
+                        mock_after_ddp_cleanup.assert_called_with(old_environ)
+
+
+@pytest.mark.parametrize("max_num_gpus_inf", [-1, 1])
 @pytest.mark.parametrize("run_extra_val_epoch", [True, False])
 @pytest.mark.parametrize("run_inference_only", [True, False])
-def test_init_inference(run_inference_only: bool, run_extra_val_epoch: bool, ml_runner_with_run_id: MLRunner) -> None:
+def test_init_inference(
+    run_inference_only: bool, run_extra_val_epoch: bool, max_num_gpus_inf: int, ml_runner_with_run_id: MLRunner
+) -> None:
     ml_runner_with_run_id.container.run_inference_only = run_inference_only
     ml_runner_with_run_id.container.run_extra_val_epoch = run_extra_val_epoch
+    ml_runner_with_run_id.container.max_num_gpus_inference = max_num_gpus_inf
+    assert ml_runner_with_run_id.container.max_num_gpus == -1
```

Comment:
Minor comment, could we add a comment here explaining that -1 is the default value for max_num_gpus? At first sight I thought it had something to do with the values passed in for max_num_gpus_inf

Reply:
Also, -1 might not be an insightful value for max_num_gpus_inf, since max_num_gpus would equal -1 both before and after init_inference got called - maybe it's worth using a different value for extra peace of mind?

Reply:
> Minor comment, could we add a comment here explaining that -1 is the default value for max_num_gpus? At first sight I thought it had something to do with the values passed in for max_num_gpus_inf

sure! I wanted to reflect that the value will change at the end of training